### ISSUE TITLE
[manuscript dev] multiple instances of viewers.ViewdocChooser

### ISSUE THREAD

**Issue by PeterBjuhr** [wbsoft/frescobaldi#760]
When debugging something else I found out that multiple instances of `viewers.ViewdocChooser` (the `QComboBox` for the manuscript files) were created.

The instances are created when calling `createWidget` in `ViewdocChooserAction`.  `createWidget` is called when the action is added to a widget. And that is done by the toolbar which it should but also by `QMenu` objects which I believe is a problem. I think this must be happening in the context menu, but couldn't find where directly.

**Comment by uliska** [comment]
I don't think it is happening in the context menu.

Actually I'm even getting *four* instances. But the result of a debug printout is

```
Create ViewdocChooser from <PyQt4.QtGui.QToolBar object at 0x7f265158c8a0>
Create ViewdocChooser from <PyQt4.QtGui.QMenu object at 0x7f265158cdf8>
about to create context menu
context menu created
Create ViewdocChooser from <PyQt4.QtGui.QMenu object at 0x7f26515559d0>
Create ViewdocChooser from <PyQt4.QtGui.QMenu object at 0x7f26515559d0>
```

So there's one widget created from the toolbar, one from a given QMenu object, *then* the context menu is created and then two ViewdocChooser widgets from the same QMenu instance.
Furthermore, if I comment out the creation of the context menu completely I'll get of course errors later on, but the three additional `ViewdocChooser`s are created nevertheless.

**Comment by PeterBjuhr** [comment]
I also did get four instances but two of them seems to be removed (garbage collected?) rather quickly.

**Comment by uliska** [comment]
I'm getting closer (to the source of the issue but also to confusion ;-) )

```
ViewdocChooser created by <PyQt4.QtGui.QToolBar object at 0x7f50e1cdc8a0>
Going to show/hide the toolbar
ViewdocChooser created by <PyQt4.QtGui.QMenu object at 0x7f50e1cdcdf8>
toolbar made visible/invisible
enter slotSessionChanged
  Session name: None
leave slotSessionChanged
ViewdocChooser created by <PyQt4.QtGui.QMenu object at 0x7fe3a39109d0>
ViewdocChooser created by <PyQt4.QtGui.QMenu object at 0x7fe3a39109d0>
enter slotSessionChanged
  Session name: 4
leave slotSessionChanged
```

* The first instance is properly created while creating the toolbar.
* The second instance is created in the `setVisible` method of the toolbar. I have isolated the complex line inside `viewers.popplerwidget.AbstractPopplerWidget.createToolbar()` to a mere `tb.setVisible(visible)` to rule out any side effects of the other calculations, and it is clear that *this* call produces the additional instance.
* The remaining two are created during startup, after a first call of changing to an empty session and before loading the actual session.

Somehow this looks like it's related to #761.
As we seem to have some fundamental issues understanding application startup - with a number of quite different symptoms - we should maybe try to seek some enlightenment in this regard. As we don't find anything in our written source code there might be some Qt things going on behind the scenes, like implicit resize events, calls to implicit methods like `createWidget()` or whatever.

---

Ah, *[EDIT]" before sending the comment: When the toolbar is set to invisible *no* additional instance is created at all

**Comment by uliska** [comment]
One more thing: I just confirmed that the four created instances are different objects (although the third and fourth have the same `QMenu` instance as parent).

**Comment by PeterBjuhr** [comment]
>When the toolbar is set to invisible *no additional instance is created at all

Ok, so something strange is happening in the toolbar...

>Somehow this looks like it's related to #761.

How do you mean exactly other than the issues happening on startup?

#761 as I see it is mainly about the first resize being so large and therefore the position is very unreliable.

**Comment by uliska** [comment]
I mean that something obscure is happening with and between a first `sessionChanged` signal and the second one where the actual session has been loaded.

More confusion:
The second instance is created when the toolbar is made visible for the first time. That is, when Frescobaldi is started with the toolbar hidden and it's shown later, then we'll get the second instance. But in that case we don't get the third and fourth instance. So these are only created when the toolbar is visible on application startup, or more concretely, between the two calls to the session slots.

**Comment by PeterBjuhr** [comment]
>I mean that something obscure is happening with and between a first sessionChanged signal and the second one where the actual session has been loaded.

Yes, but we had the resize issue before adding the explicit call to sessionChanged.

**Comment by PeterBjuhr** [comment]
I should mention that I confirmed that the music viewer has the same resizing problem so I don't think we can hope that we are doing something wrong in our viewer build if you were thinking something in that direction...

**Comment by uliska** [comment]
My main problem is that I don't have the slightest idea where these *two* executions of `app.sessionChanged` (with the first one being apparently empty) come from at application startup.

**Comment by PeterBjuhr** [comment]
The first empty one is not through the signal but from the [explicit calling](https://github.com/wbsoft/frescobaldi/blob/manuscript-viewer/frescobaldi_app/viewers/popplerwidget.py#L78). I think we had established that before...

**Comment by uliska** [comment]
Thanks for that hint, see new commit. (Doesn't help us at all but avoids unnecessary calls)

**Comment by uliska** [comment]
Some more debugging:

First instance:
```
ViewdocChooser created by <PyQt4.QtGui.QToolBar object at 0x7f15f2552938>
<viewers.ViewdocChooser object at 0x7f15f2552b00>

children of the creator
<PyQt4.QtGui.QLayout object at 0x7f15f2552c30>
<PyQt4.QtGui.QToolButton object at 0x7f15f2552cc8>
<PyQt4.QtGui.QAction object at 0x7f15f2552d60>
<PyQt4.QtGui.QWidgetAction object at 0x7f15f25529d0>
<PyQt4.QtGui.QToolButton object at 0x7f15f2552a68>
<PyQt4.QtGui.QToolButton object at 0x7f15f2552df8>
<PyQt4.QtGui.QToolButton object at 0x7f15f2552e90>
```

The parent of the `ViewdocChooser` is the toolbar, and the children (the chooser's siblings) are reported as quite generic Qt classes.

Second instance (from `setVisible(True)`:
```
ViewdocChooser created by <PyQt4.QtGui.QMenu object at 0x7f9effd8de90>
<viewers.ViewdocChooser object at 0x7f9effd8df28>

children of the creator
<PyQt4.QtGui.QAction object at 0x7f9effd520e8>
<viewers.ViewdocChooser object at 0x7f9effd8df28>
```

---

Third and fourth instances:

```
ViewdocChooser created by <PyQt4.QtGui.QMenu object at 0x7f9effd52a68>
<viewers.ViewdocChooser object at 0x7f9effd52b00>

children of the creator
<PyQt4.QtGui.QAction object at 0x7f9effd52c30>
<viewers.ViewdocChooser object at 0x7f9effd8df28>
<viewers.Zoomer object at 0x7f9effd520e8>
<PyQt4.QtGui.QSpinBox object at 0x7f9effd52180>
<viewers.ViewdocChooser object at 0x7f9effd52b00>

ViewdocChooser created by <PyQt4.QtGui.QMenu object at 0x7f9effd52a68>
<viewers.ViewdocChooser object at 0x7f9effd52d60>

children of the creator
<PyQt4.QtGui.QAction object at 0x7f9effd52e90>
<viewers.ViewdocChooser object at 0x7f9effd8df28>
<viewers.Zoomer object at 0x7f9effd520e8>
<PyQt4.QtGui.QSpinBox object at 0x7f9effd52180>
<viewers.ViewdocChooser object at 0x7f9effd52b00>
<viewers.Zoomer object at 0x7f9effd52c30>
<PyQt4.QtGui.QSpinBox object at 0x7f9effd52cc8>
<viewers.ViewdocChooser object at 0x7f9effd52d60>
```

This seems that there is a `QMenu` object somewhere that accumulates the objects from the toolbar.
In the end we have double instances of the `Zoomer` and the `QSpinBox` and even *three* of the `ViewdocChooser`!

After that we enter the session handler loading the actual session (the first, empty call has been removed by my last commit).

```
enter slotSessionChanged
Session name: 1
update view from <viewers.manuscript.ManuscriptViewdocChooserAction object at 0x7f9f00788808>
  created widget: <viewers.ViewdocChooser object at 0x7f9effd8db00>
  created widget: <viewers.ViewdocChooser object at 0x7f9effd52d60>
update view from <viewers.manuscript.ManuscriptViewdocChooserAction object at 0x7f9f00788808>
  created widget: <viewers.ViewdocChooser object at 0x7f9effd8db00>
  created widget: <viewers.ViewdocChooser object at 0x7f9effd52d60>
leave slotSEssionChanged
```

While not fully related this shows that we still call `updateViewdoc()` twice, but the `ManuscriptViewdocChooserAction` is reported to have created *two* ViewdocChooser objects.

**Comment by uliska** [comment]
> While not fully related this shows that we still call updateViewdoc() twice, but the ManuscriptViewdocChooserAction is reported to have created two ViewdocChooser objects.

Fixed in 79b1fefa

**Comment by uliska** [comment]
> Fixed in 79b1fef

*NOTE:* "Fixed" refers to the double update, not to the question of created widgets

**Comment by uliska** [comment]
I've narrowed it down even more (although I'm still puzzled exactly as before).

If you comment out the lines

`t.addWidget(self.helpButton)`
and
`self._tbAddOpenCloseActions()`
in `AbstractPopplerWidget.populateToolbar()`  only one `ViewdocChooser` instance is created. You can then even use the open/close and the help actions through the context menu.

If you *do* call the `AddOpen...` method but comment out *both* action additions it works too but if you add either of the actions we have our known behaviour again. Very strange.
I have looked through the code, but both actions don't seem to imply anything suspicious.

**Comment by uliska** [comment]
By the way: The `Zoomer` object is created four times also, identically to the `viewdocChooser`. But this is regardless of anything else in the toolbar (that is, it happens even when the zoomer is the *only* action added to the toolbar).

The `PagerAction` (which also inherits `QWidgetAction`) doesn't show this behaviour, though. It is created once, even when *all* actions are added to the toolbar.

This is very strange and somehow makes me think this issue is out of our control.

**Comment by uliska** [comment]
Some more *introspection* debug information adapted from a [stackoverflow question](http://stackoverflow.com/questions/11799290/get-function-callers-information-in-python):

```
ViewdocChooserAction created by <viewers.manuscript.ManuscriptViewPanel object at 0x7f3f1d7ae5a8>
Action object: <viewers.manuscript.ManuscriptViewdocChooserAction object at 0x7f3f1d7ae770>
```
One (single) action object created

---

```
ViewdocChooser created
My caller is the '_tbAddViewdocChooserAction' function in a 'ManuscriptViewWidget' class
Action object: <viewers.manuscript.ManuscriptViewdocChooserAction object at 0x7f3f1d7ae770>
Parent: <PyQt4.QtGui.QToolBar object at 0x7f3f1d7068a0>
Chooser object <viewers.ViewdocChooser object at 0x7f3f1d706a68>
```

This is what we'd expect: The chooser widget is created when the toolbar is populated (implicitly from `populateToolbar()`.

---


```
Going to show/hide the toolbar
ViewdocChooser created
My caller is the 'createToolbar' function in a 'ManuscriptViewWidget' class
Action object: <viewers.manuscript.ManuscriptViewdocChooserAction object at 0x7f3f1d7ae770>
Parent: <PyQt4.QtGui.QMenu object at 0x7f3f1d706df8>
Chooser object <viewers.ViewdocChooser object at 0x7f3f1d706e90>
toolbar made visible/invisible
```
This happens when `toolbar.setVisible(True)` is executed.

---

```
ViewdocChooser created
My caller is the 'widget' function in a 'ManuscriptViewPanel' class
Action object: <viewers.manuscript.ManuscriptViewdocChooserAction object at 0x7f3f1d7ae770>
Parent: <PyQt4.QtGui.QMenu object at 0x7f3f1d6c99d0>
Chooser object <viewers.ViewdocChooser object at 0x7f3f1d6c9a68>

ViewdocChooser created
My caller is the 'widget' function in a 'ManuscriptViewPanel' class
Action object: <viewers.manuscript.ManuscriptViewdocChooserAction object at 0x7f3f1d7ae770>
Parent: <PyQt4.QtGui.QMenu object at 0x7f3f1d6c99d0>
Chooser object <viewers.ViewdocChooser object at 0x7f3f1d6c9cc8>
```

Oops, it seems we have overlooked instances of the `widget()` issue. Although I don't really know what that might mean.

**Comment by PeterBjuhr** [comment]
I can't find everything wrong with *when* the widget is created now - it is created by `showEvent` in the panel base class as the other tools. But maybe there's a problem with *how* it is created...

**Comment by uliska** [comment]
I'm puzzled. I also applied that introspection to `Panel.widget()`, and it turns out that this method is executed *once*, at the beginning of the process. This confirms that it is called from the `showEvent()`, just as you say.

The second instance doesn't involve `widget()`, the output is just like in my previous comment.

The third and fourth ones now report

```
I'm ManuscriptViewdocChooserAction's createWidget()
My caller is the 'widget' function in a 'ManuscriptViewPanel' class
```
just like before. **But**, although this reports being called by `widget` in `ManuscriptViewPanel` that method isn't executed here - otherwise there would be respective output like from the first instance above.

---

The question that arises for me is if that could be a case where one single call of `widget()` produces this effect multiple times. And this leads me to the question if that could be related to that `@activate` decorator that fires a single-shot timer. What do you think?

**Comment by PeterBjuhr** [comment]
I think I managed to trace the origin to the line [`self.setWidget(w)`](https://github.com/wbsoft/frescobaldi/blob/master/frescobaldi_app/panel.py#L73). See if you can confirm that!?

**Comment by uliska** [comment]
Yes, I can. The creation of instances three and four is triggered in that line, while the creation of instances one and two is triggered in the previous line. So everything happens in the context of that first call to `widget()`. More print debugging seems to reveal the following steps:

```
showEvent()
  -> ManuscriptViewPanel.widget()
     -> createWidget()
         | - create toolbar
         | - create ViewdocChooser 1 (parent: toolbar)
          -> toolbar.setVisible() 
             # happens only  when set to True
             | - create ViewdocChooser 2 (parent: QMenu object 1)
     <---
     -> setWidget()
         # happens only when toolbar is visible
         | - create ViewdocChooser 3 (parent: QMenu object 2)
         | - create ViewdocChooser 4 (parent: QMenu object 2)
     <---
  <---
<---
```

This explains the program flow as `widget()` really is called only once (something that seemed strange so far). This also means that the delayed creation with the timer isn't part of the issue. We're not dealing with any problems of asynchronuous execution, which I somewhat was afraid of.

So my best bet is that in these events (`setVisible` and `setWidget`) there are additional obscure signals/events that we don't see and know of.
What I'm completely clueless with are these `QMenu` objects.

However, without the chance to produce a minimal example I see little chance getting help on e.g. stackexchange. And I would still suspect an issue with Frescobaldi rather than a Qt bug.

**Comment by uliska** [comment]
Maybe I've found it!
Thinking about if we had an issue with ownership/parenting here (if it *is* our fault) I read the doc for `QWidgetAction.createWidget()` for the `N`-th time.

> This function is called whenever the action is added to a container widget that supports custom widgets. If you don't want a custom widget to be used as representation of the action in the specified parent widget then 0 should be returned.

It seems our action is added to some menus (however that may be, maybe in the context of the globally managed actionCollections?). So simply checking whether the parent of the to-be-created widget is a `QMenu` suppresses the multiple creation of `ViewdocChooser` instances.

I'm not completely satisfied with this because we still don't know the origin of the issue (i.e. where these events come from. But obviously it *is* when the "action is added to a container widget"). And I'm not completely sure if the check for `QMenu` is the right one and if it shouldn't be rather an `if not isinstance` check, but for what classes?.

In any case: dce7a9d4 seems to produce correct behaviour. What do you think?

**Comment by uliska** [comment]
I also thought that we might "abuse" the QWidgetAction here.
Originally Wilbert wrote the `DocumentChooserAction` with the sole purpose of adding the widget to a toolbar. But we have added significant responsibility to it to manage the set of viewer documents.

Maybe this issue is yet another indicator that we have to rethink the separation of concerns between editor and viewer documents and the chooser widget and the underlying list management (see also #737).

What this might point to is that we need a separate class to manage the list of viewer documents (also considering future enhancements to this, like metadata-enriched manuscripts etc.) and the combobox as *one* way to display/handle them.

**Comment by PeterBjuhr** [comment]
>In any case: dce7a9d seems to produce correct behaviour. What do you think?

That seems more like a last resort when we're completely sure there's no problem with the rest of the code. And I don't think we are there yet. 

>What this might point to is that we need a separate class to manage the list of viewer documents (also considering future enhancements to this, like metadata-enriched manuscripts etc.) and the combobox as one way to display/handle them.

It's difficult to judge this when we don't know the problem, and I lack a good view of the future enhancements. But it sounds dangerously close to wanting to build a new thing instead of repairing the old and broken one.

**Comment by uliska** [comment]
Am 13.08.2015 um 15:04 schrieb Peter Bjuhr:
>     In any case: dce7a9d
>     <https://github.com/wbsoft/frescobaldi/commit/dce7a9d427f2674dba949a3ace11a92e070a46bf>
>     seems to produce correct behaviour. What do you think?
> 
> That seems more like a last resort when we're completely sure there's no
> problem with the rest of the code. And I don't think we are there yet.

I wouldn't call it a last resort, but it's very probable that it's not a
fix but a workaround.

I think the next step (before considering a PyQt bug) would be to
investigate what happens with the global actioncollection and/in the
mainwindow.
Of course it would be better to avoid the cause of the action instead of
suppressing its handler.

> 
>     What this might point to is that we need a separate class to manage
>     the list of viewer documents (also considering future enhancements
>     to this, like metadata-enriched manuscripts etc.) and the combobox
>     as one way to display/handle them.
> 
> It's difficult to judge this when we don't know the problem, and I lack
> a good view of the future enhancements. But it sounds dangerously close
> to wanting to build a new thing instead of repairing the old and broken one.

That's of course true.
But OTOH there's a small line between this and fixing a thing when it's
already clear that you need a new one anyway.
But I don't know where the truth lies in our case.

**Comment by uliska** [comment]
I did some "brute force" debugging by inserting
```python
    def createWidget(self, parent):
        if isinstance(parent, QMenu):
            parent.exec_()
```
to see what happens. And actually there *is* something happening:

* Execution is normal until the `setVisible()` call (judging from the `print` debugging)
* While the splash screen is still showing and no main window present a (context) menu pops up in the corner of the screen, containing two actions:
    * close document
    * "Select Manuscript Document"  
     (which is the text property of our viewdocChooserAction that is not displayed anywhere otherwise)
* When clicking the second option twice execution continues normally but only a third viewdocchooser is created
* When clicking the first option behaviour is somewhat unclear but doesn't really help understanding the thing

So obviously there *is* a Menu somewhere, collecting some of our actions but normally never executed.

---

Another thing I tested is what the *parent* of that menu might be, and it turns out that it's a `QToolButton` instance.

What puzzles me is that the three `QMenu` instances have three different `QToolButton` instances as parents while two of the `QMenu` instances are supposed to be the same objects.

Calling `parent.parent().showMenu()` instead of the command at the top of this comment shows the same popup menu but not in the corner of the screen but at the place where the help  and the close buttons in the toolbar will show up later.
So it becomes clear that this obscure `QMenu` instances belong to the buttons of the help and the close manuscript actions. Still it's strange that to avoid the wrong behaviour you have to comment out the help, the close *and* the open actions when populating the toolbar.

`QToolButton` has an implicit `menu()` property.
Is a `QToolButton` added implicitly to an action when it's added to a toolbar?
Is a context menu implicitly added to the toolbar in our case as suggested by this doc?

> **QWidget.addAction (self, QAction action)**  
Appends the action action to this widget's list of actions.  
All QWidgets have a list of QActions, however they can be represented graphically in many different ways. The default use of the QAction list (as returned by actions()) is to create a context menu.

But then again it wouldn't make sense that exactly *these* actions are part of the context menu and not the others.

**Comment by uliska** [comment]
> Is a QToolButton added implicitly to an action when it's added to a toolbar?

RTFM:
> Tool buttons are normally created when new QAction instances are created with QToolBar.addAction() or existing actions are added to a toolbar with QToolBar.addAction(). 

So this *may* be an explanation where our dubious `QMenu` instances come from - being created implicitly as part of the `QToolButtons` implicilty created through `addAction()`. However, this doesn't explain why only these few actions have that behaviour and cause the chooser to be created.

**Comment by uliska** [comment]
Just for the record (and stopping for this evening):
I have reimplemented the help button as a regular `QAction`. Doesn't change *anything*, nevertheless I'm sure it's better coding now.

**Comment by PeterBjuhr** [comment]
Yes, I also thought about that!

**Comment by PeterBjuhr** [comment]
It turns out that if we add/place the documentchooser first in the toolbar the problem is "solved". Although I think we can live with that little rearrangement, I really would like to know why (is it a Qt bug?)! Perhaps it is possible to create a minimal example after all...!

**Comment by uliska** [comment]
Huh???

Strange ...
-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by PeterBjuhr** [comment]
I can't reproduce this in the Musicviewer so it seems we will find the bug in the viewer/manuscript code after all!

**Comment by uliska** [comment]
> It turns out that if we add/place the documentchooser first in the
> toolbar the problem is "solved". Although I think we can live with that
> little rearrangement, 

What rearrangement would you suggest? The only one I can imagine would be moving the help button to the right end of the toolbar.
But I'm not so sure I would like to "live" with that.

Right now we assume we have a bug somewhere, be it in our coding or in Qt. And we hope to fix that one day, sooner or later. When we apply a workaround we assume this to be temporary.
It seems we have a choice between rearranging the buttons and suppressing adding the chooser to the toolbuttons' menus. Both are workarounds and no cures.

I have two points for "my" approach:
* It doesn't affect the visible appearance, a later proper fix won't change the UI
* It's somewhat in line with the documentation I quoted in an earlier comment:
    * We know that the widget is added (for a reason we do *not* know) to the menus attached to some of the toolbuttons. And the documentation suggests to return `None` if we don't want that to happen. So suppressing the event like I did in dce7a9d4 seems *quite* canonical.

>  I really would like to know why (is it a Qt bug?)! Perhaps it is possible to create a minimal example after all...!

That would be good, although I can't see how yet.

> I can't reproduce this in the Musicviewer so it seems we will find the bug in the viewer/manuscript code after all!

So, what are the differences between the old and the new viewer?

* The Musicview doesn't have the toolbar integrated but only in the mainwindow
* The Musicview's toolbar doesn't have the help button nor the add/close actions - so the probable culprits aren't there.

**Comment by PeterBjuhr** [comment]
>>It turns out that if we add/place the documentchooser first in the
toolbar the problem is "solved". Although I think we can live with that
little rearrangement,

>What rearrangement would you suggest? 

What I meant was that having the documentchooser as the leftmost action isn't that bad if it would help anything. I'm far from convinced that we should do it so I'm not suggesting it!

>The only one I can imagine would be moving the help button to the right end of the toolbar. But I'm not so sure I would like to "live" with that.

Actually I would prefer this move, I'll get back to this...

And moving the documentchooser to the left would make the toolbar more consistent with the music view toolbar. So actually there are advantages to the rearrangement. But I think having the open and close buttons to the left is more intuitive. And as a solution to the problem it is obviously totally sweeping the issue under the rug.

>Right now we assume we have a bug somewhere, be it in our coding or in Qt. And we hope to fix that one day, sooner or later. When we apply a workaround we assume this to be temporary.
It seems we have a choice between rearranging the buttons and suppressing adding the chooser to the toolbuttons' menus. Both are workarounds and no cures.

I agree. And I haven't noticed any real problems so it may be a bug we can live with and hope it would one day be solved (maybe by a qt upgrade or by a redesign we are doing anyway). What I mean is that we may not actually need a temporary workaround. But what I would like before closing this is to make sure we are not doing something wrong ourselves. 

>>I can't reproduce this in the Musicviewer so it seems we will find the bug in the viewer/manuscript code after all!

>So, what are the differences between the old and the new viewer?

>The Musicview doesn't have the toolbar integrated but only in the mainwindow.

Do you think it matters where the toolbar is placed? Maybe we could debug moving the toolbar and see if it makes any difference!?

>The Musicview's toolbar doesn't have the help button nor the add/close actions - so the probable culprits aren't there.

In my test I put the print action before the documentchooser...

I may be wrong but I think that if it really is a qt bug we should be able to reproduce it also in the music viewer (somehow...). Not being able to reproduce it strongly suggests to me that the problem lies in our coding.

**Comment by uliska** [comment]
> And moving the documentchooser to the left would make the toolbar more
> consistent with the music view toolbar. 

I wouldn't really follow you here. Consistency with the music view
toolbar isn't an issue because we will adapt that one to our new
structure later. So any inconsistency will only be temporary, and I'd
prefer this temporary state over changing the layout of the new toolbar
again in the near future.

> So actually there are advantages
> to the rearrangement. But I think having the open and close buttons to
> the left is more intuitive. 

Yes.

> And as a solution to the problem it is
> obviously totally sweeping the issue under the rug.

Yes ;-)

>     Right now we assume we have a bug somewhere, be it in our coding or
>     in Qt. And we hope to fix that one day, sooner or later. When we
>     apply a workaround we assume this to be temporary.
>     It seems we have a choice between rearranging the buttons and
>     suppressing adding the chooser to the toolbuttons' menus. Both are
>     workarounds and no cures.
> 
> I agree. And I haven't noticed any real problems so it may be a bug we
> can live with and hope it would one day be solved (maybe by a qt upgrade
> or by a redesign we are doing anyway). What I mean is that we may not
> actually need a temporary workaround. But what I would like before
> closing this is to make sure we are not doing something wrong ourselves.

So I suggest we keep the behaviour introduced in
dce7a9d427f2674dba949a3ace11a92e070a46bf as a temporary solution but add
some explicit source comments and add a new issue for it. (I think
that's better than simply keeping this one open. Reference the old issue
and put in the initial comment of the new issue what we have distilled
as information so far.)

> 
>         I can't reproduce this in the Musicviewer so it seems we will
>         find the bug in the viewer/manuscript code after all!
> 
>     So, what are the differences between the old and the new viewer?
> 
>     The Musicview doesn't have the toolbar integrated but only in the
>     mainwindow.
> 
> Do you think it matters where the toolbar is placed? Maybe we could
> debug moving the toolbar and see if it makes any difference!?

No, I don't think it matters. But as we are pretty clueless it should be
considered too.

> 
>     The Musicview's toolbar doesn't have the help button nor the
>     add/close actions - so the probable culprits aren't there.
> 
> In my test I put the print action before the documentchooser...

Well, the print action doesn't do any harm in the manuscript viewer
either. If you remove the help, add and close actions and put the print
action before the documentchooser everything is fine there too, isn't it?

> 
> I may be wrong but I think that if it really is a qt bug we should be
> able to reproduce it also in the music viewer (somehow...). Not being
> able to reproduce it strongly suggests to me that the problem lies in
> our coding.

Yes and no.
Even if it is a Qt bug it must be triggered by some special
circumstances, and these presumably lie in our coding.

Anyway, we should definitely try to find out what is wrong there. But
maybe it doesn't have to block opening a pull request.
Having several instances created might seem neglectable, but I don't
think it is.
- it's not only the documentchooser but also the zoomer (I think)
- it will probably affect a number of new-style viewers
- we don't have an idea about potential side-effects:
  what if at some point something is accessing the wrong instance,
  for example?

So I think my above suggestion should be the way to go:
- keep my workaround with suppressing the creation for the ToolButtons
- Move the help button, but independently
- create a new issue with a summary of our research.

**Comment by uliska** [comment]
One more thing: Right now my workaround is implemented in `ViewdocChooserAction.createWidget()` but I think I have to move it to the `ComboBoxAction` class because it also affects the Zoomer. (Check if it's really (and only) the Zoomer that is affected).

**Comment by PeterBjuhr** [comment]
Ok, but let's first make one more effort to understand the issue...

**Comment by PeterBjuhr** [comment]
>>In my test I put the print action before the documentchooser...

>Well, the print action doesn't do any harm in the manuscript viewer
either. If you remove the help, add and close actions and put the print
action before the documentchooser everything is fine there too, isn't it?

No, it isn't. I just retried it. Did you forget to remove your fix before testing?

**Comment by PeterBjuhr** [comment]
>Anyway, we should definitely try to find out what is wrong there. But
maybe it doesn't have to block opening a pull request.

Agreed.

>Having several instances created might seem neglectable, but I don't
think it is.
>- it's not only the documentchooser but also the zoomer (I think)
>- it will probably affect a number of new-style viewers
>- we don't have an idea about potential side-effects:
>  what if at some point something is accessing the wrong instance,
  for example?

Admittingly the potential side-effects are difficult to grasp. 

>So I think my above suggestion should be the way to go:
>- keep my workaround with suppressing the creation for the ToolButtons

My only concern with this (other than it being premature to create a workaround before we knew if there is a real solution) is that we forget about it. So a comment in the code would be good.

>- Move the help button, but independently.

Yes, this is independent at this point.

>- create a new issue with a summary of our research.

I don't see why it is helpful to create a new issue. I think this is quite spot on. But I have no major objection.

**Comment by uliska** [comment]
>>>  In my test I put the print action before the documentchooser...

>> Well, the print action doesn't do any harm in the manuscript viewer
    either. If you remove the help, add and close actions and put the print
    action before the documentchooser everything is fine there too, isn't it?

> No, it isn't. I just retried it. Did you forget to remove your fix before testing?

No, but I didn't realize how to properly test it. All I did was commenting and uncommenting actions, but I didn't think of the order.
Now I've put the documentchooser at the very end of the list and commented everything else out. As soon as *any* action is added before the chooser the behaviour appears, so it is *not* related at all to the ones I "identified" earlier.

So to sum up (as it is possible now):

* When a `QAction` is added to a `QToolBar` it implicitly creates a `QToolButton`
* This toolbutton implicitly creates a `QMenu`
* When a `QWidgetAction` descendant is added to the toolbar afterwards is is implicitly (erronously?) added to the menu of a (actually: which one?) toolbutton.

But this last one doesn't happen upon creation but *once*  when `setVisible(True)` is executed and again *twice* when `setWidget()` is called.

**Comment by uliska** [comment]
While this issue *is* quite spot on it is too lengthy to act as a primary reference for future investigation of this issue. There are so many detours here that it will be hard to quickly see what's relevant. Therefore I created #762 as the new issue, which isn't tied to any milestone.

**Comment by PeterBjuhr** [comment]
If the other issue is for future reference we can keep the current discussion here!??

**Comment by uliska** [comment]
I think as long as it's "current" discussion, yes, we can continue here. The new issue should only be updated if we come to significant new insights. Once we stop thinking about this issue *with the perspective of our initial release*  there should be the new issue as a "clean start".

**Comment by PeterBjuhr** [comment]
Alright, a future safe approach! :)

**Comment by PeterBjuhr** [comment]
>>So, what are the differences between the old and the new viewer? The Musicview doesn't have the toolbar integrated but only in the mainwindow.

 >Do you think it matters where the toolbar is placed? Maybe we could
 debug moving the toolbar and see if it makes any difference!?

That's it!!! Testing confirms this idea! 

I became more convinced after reading about limitations when the toolbar isn't a child of the mainwindow:

>When a QToolBar is not a child of a QMainWindow, it loses the ability to populate the extension pop up with widgets added to the toolbar using addWidget(). Please use widget actions created by inheriting QWidgetAction and implementing QWidgetAction::createWidget() instead.

**Comment by PeterBjuhr** [comment]
Interesting, even when the toolbar is a child of the mainwindow you can reproduce the issue by moving the toolbar off the mainwindow!

![selection_011](https://cloud.githubusercontent.com/assets/355116/9284831/b5d7fa28-42df-11e5-887b-9301f166844e.png)

**Comment by uliska** [comment]
So what do you mean exactly? Am I right if I understand the following:

* make the toolbar a child of the mainwindow instead of the viewer widget/panel makes our problem go away
* But if you rip it off and make it floating it's back again?

**Comment by PeterBjuhr** [comment]
Yes. As long as the toolbar stays on the mainwindow everything is fine. But when the toolbar is floating the extra widgets are created by the superfluous menu instances as we have seen.

**Comment by uliska** [comment]
Very interesting indeed.
But what does that mean for us? That we can be more content with my
workaround or that it will be harder to find a real cure (i.e. an
approach to things that doesn't trigger this weird behaviour)?

**Comment by PeterBjuhr** [comment]
I don't see everything clear yet. But here are some thoughts:

I guess the safest thing would be to place the toolbar in the mainwindow. We don't know if there are other side-effects that we haven't discovered yet (of not having the toolbar in the mainwindow). But I can't now see any design solution that would not be very awkward.

Your workaround seems quite sound; I can't imagine wanting to add the widgetactions to a menu anyway. But as the viewers are so important I feel that we should perhaps not settle with a quick fix. But the alternative as I see it now is to implement our own toolbar class (either as a child of QWidget or QToolBar).

**Comment by uliska** [comment]
> I guess the safest thing would be to place the toolbar in the mainwindow

You mean "physically" moving from the viewer to the mainwindow's toolbar?
I admit I have strong difficulties imagining that to work when we have multiple more or less equal toolbars side by side.

> as the viewers are so important I feel that we should perhaps not settle with a quick fix.

I fully agree with that.

> implement our own toolbar class (either as a child of QWidget or QToolBar).

You have an idea how to manage that in order to overcome the problematic behaviour?

But I might be interested in starting this. First I'd try to make a class that doesn't change anything (i.e. uses an unmodified QToolBar descendant). But I'd make it so that the layout with the help button is built-in to the toolbar class.

**Comment by PeterBjuhr** [comment]
>>I guess the safest thing would be to place the toolbar in the mainwindow

>You mean "physically" moving from the viewer to the mainwindow's toolbar?
I admit I have strong difficulties imagining that to work when we have multiple more or less equal toolbars side by side.

In theory the toolbar can be placed anywhere as long as it is directly on the mainwindow and not on some other widget. So it is not unthinkable to find a design solution where the toolbar have a suitable place in connection to its viewer. But having tried to move the toolbar around I would say that the diffulties are rather huge. 

>>as the viewers are so important I feel that we should perhaps not settle with a quick fix.

>I fully agree with that.

>implement our own toolbar class (either as a child of QWidget or QToolBar).
You have an idea how to manage that in order to overcome the problematic behaviour?

Not much, but I'm thinking either build up the toolbar functionality from a QWidget or override the problematic aspects of QToolBar itself.

**Comment by uliska** [comment]
See 36722026. If we want to give your idea a try this is a good start. And if we decide not to go that route it is probably still a good thing for code organization.

**Comment by uliska** [comment]
There is an empty `ToolBar` class [here](https://github.com/wbsoft/frescobaldi/blob/manuscript-viewer/frescobaldi_app/viewers/toolbar.py#L33). This is used for the two inner toolbars. If you want to try building from `QWidget` this would be simply replaced here.

**Comment by uliska** [comment]
In this `ToolBar` class we could override `addAction(self, action)`.
Do you think it could be useful to take the passed action, manually create a `QToolButton` (if it's not a `QWidgetAction` and manually add that toolbutton to the toolbar? It might be possible to do that in a way so the widgetaction is not added to the new toolbutton's menu.

See 374e6201

**Comment by uliska** [comment]
... and 80680a65

**Comment by PeterBjuhr** [comment]
>One more thing: Right now my workaround is implemented in ViewdocChooserAction.createWidget() but I think I have to move it to the ComboBoxAction class because it also affects the Zoomer. (Check if it's really (and only) the Zoomer that is affected).

Although we are moving away from the workaround I thought I should mention this: I just realized a major drawback - because all `QWidgetAction`:s are affected that includes `PagerAction` which is not a child of `ComboBoxAction` and therefore out of reach of the fix even after the move.

**Comment by uliska** [comment]
>     One more thing: Right now my workaround is implemented in
>     ViewdocChooserAction.createWidget() but I think I have to move it to
>     the ComboBoxAction class because it also affects the Zoomer. (Check
>     if it's really (and only) the Zoomer that is affected).
> 
> Although we are moving away from the workaround I thought I should
> mention this: I just realized a major drawback - because all
> |QWidgetAction|:s are affected that includes |PagerAction| which is not
> a child of |ComboBoxAction| and therefore out of reach of the fix even
> after the move.

Ah yes, that's of course true.
A rather trivial fix for that would be a common base class inheriting
from QWidgetAction where all our widget actions inherit from.
My question is: is it more appropriate to implement a fix by overriding
some QToolBar behaviour (so it properly adds actions and widget actions)
or by overriding some QWidgetAction behaviour (so it is properly added
to a toolbar)?

Somehow I have the impression that reimplementing the toolbar itself
isn't "cleaner" in any way than what I suggested.

**Comment by PeterBjuhr** [comment]
I think it is cleaner to write an own toolbar class that does what we want instead of overlaying workarounds to behavior of the original toolbar that is unintended or at least unwanted. But it's a more complex task and may require more extensive code changes.

**Comment by PeterBjuhr** [comment]
If I may add to the last comment: Basically what you're comparing is a overrides of QToolBar or QWidgetAction. And in principle it seems strange to me to override QWidgetAction when it is QToolbar that has the unwanted behavior.

**Comment by uliska** [comment]
Am 16.08.2015 um 12:25 schrieb Peter Bjuhr:
> If I may add to the last comment: Basically what you're comparing is a
> overrides of QToolBar or QWidgetAction. And in principle it seems
> strange to me to override QWidgetAction when it is QToolbar that has the
> unwanted behavior.

If it *is* unambiguously clear that it really is the QToolBar

**Comment by PeterBjuhr** [comment]
Yes, that basic assumption can still turn out to be false. And rebuilding QToolbar can turn out to be too complex for our current purposes. But I answered the question given what we know now.

**Comment by uliska** [comment]
I think I've prepared the infrastructure for that by factoring out the toolbar and with what I srarted on the custom-toolbar branch. First tests with overriding addAction worked without problems - but also without improvements.

**Comment by uliska** [comment]
BTW: I may be mistaken and I don't know why, but I have the impression the "second" instance has vanished, there are only the two instances 3 and 4 created by now.

**Comment by PeterBjuhr** [comment]
Are you referring to the custom-toolbar branch?

**Comment by uliska** [comment]
I did so, but now (3fb0c32e) it seems that *no* additional widgets are created anymore. Of course at the prize of oru widgets not working properly ;-)

But it seems to confirm that the culprit is somewhere inside `addAction()`

**Comment by PeterBjuhr** [comment]
No, I don't think we can draw that conclusion! See my latest comment in that commit. The actions are turned into regular QWidgetAction:s and we can't debug what happens anymore.

Let me explain that further: when we call addWidget a new QWidgetAction is created (if my reading of the source is correct). And the widget is connected to that instead of the correct QWidgetAction.

**Comment by uliska** [comment]
Maybe I'm too stupid but I didn't find the source code. Could you give me a pointer?

**Comment by PeterBjuhr** [comment]
Yes, of course! This is what I studied: http://cep.xray.aps.anl.gov/software/qt4-x11-4.2.2-browser/d1/d58/qtoolbar_8cpp-source.html.

**Comment by uliska** [comment]
So it seems to me like we should figure out what [`QWidget.addAction`](http://cep.xray.aps.anl.gov/software/qt4-x11-4.2.2-browser/db/d0a/class_q_widget.html#8919f8e3aacb45a7d5f7fac2d226c0d8)` internally does.
From the description it's not really clear to me how it will (in `QToolBar`) decide to add a tool button to the toolbar.

Maybe we could go back one more step and try the following:

* Inherit directly from `QWidget`
* Add a layout
* in `addAction` (for `QAction`) create a `QToolButton` and add that to the layout
* in `addAction` (for `QWidgetAction`) create the widget through `createWidget()` and add *that* to the layout.

However, I don't know how that will work out with the *action* part of our object. Somehow the action would have to be added to the widget (like to the toolbutton), but I don't really see if/that this would easily work.

**Comment by PeterBjuhr** [comment]
>So it seems to me like we should figure out what QWidget.addActioninternally does. From the description it's not really clear to me how it will (inQToolBar`) decide to add a tool button to the toolbar.

That part seems to be done in `createItem`.

>Maybe we could go back one more step and try the following:

>Inherit directly from QWidget
>* Add a layout
>* in addAction (for QAction) create a QToolButton and add that to the layout
>* in addAction (for QWidgetAction) create the widget through createWidget() and add that to the layout.

I'm thinking in the same direction (and have actually started testing...) It seems to complex to try to override the behavior in QToolBar.

>However, I don't know how that will work out with the action part of our object. Somehow the action would have to be added to the widget (like to the toolbutton), but I don't really see if/that this would easily work.

Except `setDefaultAction` for the widget there is also the corresponding `setDefaultWidget` for the action.

But I have now idea what it takes to get it too work...

**Comment by PeterBjuhr** [comment]
For some reason when I do `action.setDefaultWidget` the comboxes disappear. 

But I can push so you'll get a chance to take a look!?

**Comment by uliska** [comment]
Good idea (although I'm not sure about my time tonight ). Maybe on a new branch

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by PeterBjuhr** [comment]
Ok, I created a new branch [qwidget-toolbar](https://github.com/wbsoft/frescobaldi/tree/qwidget-toolbar).

**Comment by uliska** [comment]
Thanks.
It's too late now to really investigate that but I have to say it somewhat looks circular to set the widget to an action that is created by a widget.
Hope to find time tomorrow.

**Comment by PeterBjuhr** [comment]
>Interesting, even when the toolbar is a child of the mainwindow you can reproduce the issue by moving the toolbar off the mainwindow!

Using the method above I thought I'd try to reproduce the error in the musicview's toolbar. But I couldn't!

It's getting more and more clear to me that the error lies somewhere in our own code and that Qt is not to be blamed. 

While trying out the "home made" toolbar I also encountered a strange error that looks like it could be related. And although I can't rule out a Qt bug in that case it seems to me somewhat unlikely. 

A short description is that instead of strange items being added to the widgets of the widgetactions the strange items are replacing the correct widgets.

**Comment by uliska** [comment]
> It's getting more and more clear to me that the error lies somewhere in our own code and that Qt is not to be blamed. 

What makes me wonder is that also the pager is affected, and we didn't touch that at all, isn't it?

> A short description is that instead of strange items being added to the widgets of the widgetactions the strange items are replacing the correct widgets.

Sorry, but this is *too* short for me to get an idea.

**Comment by uliska** [comment]
> It's getting more and more clear to me that the error lies somewhere in our own code and that Qt is not to be blamed. 

Maybe this is related to what I expressed in #737. Maybe we're not using the `QWidgetAction` properly after all, but I'm not sure as its description in the docs seems to indicate exactly our use case. But it turns out now that Wilbert's original design also is not completely consistent. For example: The musicview's toolbar is not completely tied to the viewer but lives independently in the mainwindow's toolbar. If you close the musicview tool the toolbar is still visible, and if you do something in the toolbar then (e.g. change zoom) the viewer will be made visible again.

What is it that we actually have and want?

* A ComboBox in the viewer's toolbar
* This holds the names of the currently available viewer documents
* When a different item in the combobox is selected the corresponding document should be displayed
* The items can be populated through different means, explicitly via the file open dialog or implicitly be reacting to a change in the editor documents

It seems this should be doable by a plain combobox and an object that implements the functionality, i.e. slots for the combobox signals and the other interesting signals. In fact it should more or less implement the same as our current `ViewdocChooserAction` but inherit from `QAction` or even `QObject`. Then we could implement the parts from `QWidgetAction` that are necessary to tie that to the combobox, which is then added through `addWidget` to our `QWidget` based Toolbar.

Of course this is some work and would trigger a new round of development *and testing*, but it could well make the relations clearer to us, and would help us avoid working around some things. For example the stuff in `ViewdocChooser` might as well be obsolete because after all we don't want our widget to be instantiated several times in several parents.
One more thing to consider of course is the relation to the pager widget. Of course we'd need a common base class for these.

Does that make any sense?

**Comment by PeterBjuhr** [comment]
>>A short description is that instead of strange items being added to the widgets of the widgetactions the strange items are replacing the correct widgets.

>Sorry, but this is too short for me to get an idea.

Sorry, I'll try another approach still simplistic:

Before we had multiple instances of the widgets, now we get the wrong widgets, e. g. a ToolButton instead of a combobox.

**Comment by PeterBjuhr** [comment]
>It seems this should be doable by a plain combobox and an object that implements the functionality, i.e. slots for the combobox signals and the other interesting signals. In fact it should more or less implement the same as our current ViewdocChooserAction but inherit from QAction or even QObject. Then we could implement the parts from QWidgetAction that are necessary to tie that to the combobox, which is then added through addWidget to our QWidget based Toolbar.

I don't think we should try to redesign or restructure anything at this point. We just have to keep debugging until we find what goes wrong. As we don't have been able to reproduce the issue in the music viewer I think we should carefully scrutinize the differences and see it we can spot anything.

**Comment by uliska** [comment]
One thing where we have introduced rather big changes is what is in the widget and what is in the panel classes.
But that doesn't explain why also the pager is affected.

**Comment by PeterBjuhr** [comment]
>One thing where we have introduced rather big changes is what is in the widget and what is in the panel classes.

Yes, we should take a look at that!

>But that doesn't explain why also the pager is affected.

Because we didn't change it? It seems all WidgetActions are affected. If we want we could try to confirm that by creating an additional test action of as a pure QWidgetAction.

**Comment by uliska** [comment]
This is a good idea. But if we didn't change it and it is still affected, doesn't that indicate that the error is *not* in our code?

**Comment by PeterBjuhr** [comment]
At least it indicates that the error is not in that code. And that was what I meant - because all widgetactions are affected we will not find the error in any particular implementation.
